src/cargo/core/compiler/output_sbom.rs

```diff
@@ -0,0 +1,292 @@
+pub struct SbomFormatVersion<const V: u32>;
+
+impl<const V: u32> Serialize for SbomFormatVersion<V> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_u32(V)
+    }
+}
+
+/// A profile can be overriden for individual packages.
+///
+/// This wraps a [`Profile`] object.
+/// See <https://doc.rust-lang.org/nightly/cargo/reference/profiles.html#overrides>
+#[derive(Serialize, Clone, Debug)]
+struct SbomProfile {
+    name: String,
+    opt_level: String,
+    lto: Lto,
+    codegen_backend: Option<String>,
+    codegen_units: Option<u32>,
+    debuginfo: DebugInfo,
+    split_debuginfo: Option<String>,
+    debug_assertions: bool,
+    overflow_checks: bool,
+    rpath: bool,
+    incremental: bool,
+    panic: String,
+    #[serde(skip_serializing_if = "Vec::is_empty")] // remove when `rustflags` is stablized
+    rustflags: Vec<String>,
+}
+
+impl From<&Profile> for SbomProfile {
+    fn from(profile: &Profile) -> Self {
+        let rustflags = profile
+            .rustflags
+            .iter()
+            .map(|x| x.to_string())
+            .collect_vec();
+
+        Self {
+            name: profile.name.to_string(),
+            opt_level: profile.opt_level.to_string(),
+            lto: profile.lto,
+            codegen_backend: profile.codegen_backend.map(|x| x.to_string()),
+            codegen_units: profile.codegen_units.clone(),
+            debuginfo: profile.debuginfo.clone(),
+            split_debuginfo: profile.split_debuginfo.map(|x| x.to_string()),
+            debug_assertions: profile.debug_assertions,
+            overflow_checks: profile.overflow_checks,
+            rpath: profile.rpath,
+            incremental: profile.incremental,
+            panic: profile.panic.to_string(),
+            rustflags,
+        }
+    }
+}
+
+/// Describes a package dependency
+#[derive(Serialize, Clone, Debug)]
+struct SbomPackage {
+    package_id: PackageIdSpec,
+    package: String,
+    /// Package profile is given when it differs from the root level [`Profile`].
+    #[serde(skip_serializing_if = "Option::is_none")]
+    profile: Option<SbomProfile>,
+    version: Option<Version>,
+    features: Vec<String>,
+    build_type: SbomBuildType,
+    extern_crate_name: String,
+    /// Indices into the `packages` array
+    dependencies: Vec<usize>,
+}
+
+impl SbomPackage {
+    pub fn new(unit_dep: &UnitDep, root_profile: &Profile) -> Self {
+        let package_id = unit_dep.unit.pkg.package_id().to_spec();
+        let package = package_id.name().to_string();
+        let profile = if &unit_dep.unit.profile != root_profile {
+            Some((&unit_dep.unit.profile).into())
+        } else {
+            None
+        };
+        let build_type = if unit_dep.unit.mode.is_run_custom_build() {
+            SbomBuildType::Build
+        } else {
+            SbomBuildType::Normal
+        };
```

Comment:
The documentation for `is_run_custom_build` says
```rust
    /// Returns `true` if this is the *execution* of a `build.rs` script.
```

So this isn't saying whether this is a normal or build instance but whether this will be run as a build script.

Reply:
as in build scripts show up twice, once for the Build mode and once for the Run mode